### DISCUSSION TITLE
Self-Hosted Solution

### DISCUSSION THREAD

**Discussion by abutlerboudakian** [ThioJoe/YT-Spammer-Purge#40] [Ideas]
Are there any plans at the moment to make this a self-hosted webapp (i.e. purchasing a domain and hosting a web version of this)? Having this tool available will be great for weeding out spammers but a lot of youtubers might not want to go through all the effort of generating their own API keys and running the console application manually, as well as possibly getting sussed out by all the security warnings that come up when downloading the compiled exe.

**Comment by ThioJoe** [comment]
I thought about it, and it definitely would be way easier to use, but I'd have to do a lot of research to figure out how to do that. Also the YouTube API has a relatively low initial quota and I'm not sure how hard it would be to get it increased sufficiently. I'll look into it, but don't really have any plans for it at the moment.

**Comment by arjun-dureja** [comment]
FYI, I've gotten my YouTube API quota increased from 25K per month to 500K per month pretty easily. I just had to submit a request explaining what my app did (mine was a simple sub count tracker) and why I needed a higher quota. They took about 3 weeks to get back to me.

**Comment by joaogelado** [comment]
> I thought about it, and it definitely would be way easier to use, but I'd have to do a lot of research to figure out how to do that. Also the YouTube API has a relatively low initial quota and I'm not sure how hard it would be to get it increased sufficiently. I'll look into it, but don't really have any plans for it at the moment.

If you do so, I recommend using vercel. It's like plug and play and it has a free plan that is good enough for starter projects. Also I like Next js (that is also from Vercel), but for this you will need learn JavaScript. Anyways, this is not a paid comment, it is just my opinion as a developer. Also, I am a programmer and use js for my projects. If you think I may help with something, you can try to call me out.

Also, if I can make one, I will share my example here so you can see it.

**Comment by ArjixWasTaken** [comment]
> FYI, I've gotten my YouTube API quota increased from 25K per month to 500K per month pretty easily. I just had to submit a request explaining what my app did (mine was a simple sub count tracker) and why I needed a higher quota. They took about 3 weeks to get back to me.

Easy solution is to use multiple API keys from multiple google cloud console projects.

**Comment by deidyomega** [comment]
If you decide you want to go down that route, I'd be happy to help you do it. I've built more than a few webapps.

**Comment by SevastianGit** [comment]
> Easy solution is to use multiple API keys from multiple google cloud console projects.

Hmm, if that works and does increase the quota, maybe we should make a new issue to request this as a feature?

**Comment by TechStudent10** [comment]
@ThioJoe It's actually pretty easy when you think about it. **At first.** Since you are using Python, using something like Flask would work. All you would have to do is figure out some way to manage users AND how to send them a sign-in link. What I mean in the first one (just to avoid confusion) is how are you going to distinguish different users? If you use something like [Flask-Dance](https://flask-dance.readthedocs.io/en/latest/index.html), then yeah it's possible. **However**, Google, for whatever reason, decided to go out and make their own library instead of just allowing us to get a simple Client ID and Secret and then allowing us to make requests to an API where you can delete the comments. I guess it's for security purposes...

**Reply by TechStudent10** [reply]
If I can find a way to make it work, I'll submit a pull request. But I would have to do some research...

**Reply by TechStudent10** [reply]
This might be relevent: https://realpython.com/flask-google-login/.

**Reply by abutlerboudakian** [reply]
When I suggested this issue if Python was to be the language still I was thinking Django would probably make more sense, as that allows for easy front-end implementation using SSR templating, as well as easy OAuth with something like AllAuth/DJRestAuth

**Reply by TechStudent10** [reply]
Flask-Dance is also very easy to implement. Plus, I don't think ThioJoe wants to basically rewrite the program from the bottom up to work with Django. Flask is easier to use and implement. He could literally copy and paste code. But like I said, user management is still a problem...

**Reply by abutlerboudakian** [reply]
Not asking ThioJoe to rewrite anything, if you look at my fork I spent a little time after work today throwing together a basic implementation, it's not even remotely close to done yet but it should be prototype-able with a few more hours of work. Even if it doesn't end up getting picked up by him or hosted at least the code will be public in case anyone attempts a better implementation than mine in the future.

**Reply by TechStudent10** [reply]
@abutlerboudakian have you figured out any of the issues? Like distinguishing users? Or sending a sign-in link? Not trying to be mean, just asking.

**Reply by reesericci** [reply]
Honestly, I find something like Koa + Node.js to be more suitable because of the framework that is Koa/Express. It's really easy to build on top of, although that would require a total rewrite.

**Reply by abutlerboudakian** [reply]
@TechStudent11  That stuff's already implemented in the fork, it's not pretty at the moment but it is functional. The package I'm using is called django-google-api, which generates a migration for a GoogleAuth model which keeps track of each token that gets generated through a redirect-uri after authenticating. Right now I only have localhost set up as the callback-url since I'm testing it for hosting yet, but it can autheticate and return tokens correctly.

**Reply by abutlerboudakian** [reply]
Here's the snippet that's doing most of the work:

```
from django_google.flow import DjangoFlow, CLIENT_SECRET_FILE, SCOPES
from django_google.models import GoogleAuth
from django.contrib.auth import get_user_model

def not_broken_constructor(self, oauth2session, client_type, client_config, redirect_uri=None, code_verifier=None, autogenerate_code_verifier=None):
    os.environ['OAUTHLIB_RELAX_TOKEN_SCOPE'] = '1'
    super(DjangoFlow, self).__init__(oauth2session, client_type, client_config,  redirect_uri, code_verifier, autogenerate_code_verifier)

DjangoFlow.__init__ = not_broken_constructor

User = get_user_model()
flow = DjangoFlow.from_client_secrets_file(client_secrets_file=CLIENT_SECRET_FILE, scopes=SCOPES)

def oAuthView(request):
        callback_url=reverse("oauth2callback") # callback Url (oAuth2CallBackView URL)
        return redirect(flow.get_auth_url(request, callback_url=callback_url))

# Google Authentication Call Back VIEW (Using Without Javascript)
def oAuth2CallBackView(request):
    success_url = "/dashboard/"  # redirection URL on Success reverse() can b use here

    # Library a little sus if the public api has mispelled methods
    creds = flow.get_credentails_from_response_url(response_url=request.build_absolute_uri())
    
    userinfo = flow.get_userinfo(creds=creds)
    try:
        user = User.objects.get(email=userinfo['email'])
    except User.DoesNotExist:
        user = User.objects.create(
            email=userinfo['email'],
            username=userinfo['email'],
            first_name=userinfo['given_name'],
            last_name=userinfo['family_name']
        )
    finally:
        try:
            gauth = GoogleAuth.objects.get(user=user)
        except GoogleAuth.DoesNotExist:
            gauth = GoogleAuth.objects.create(user=user, creds=creds)
```

**Reply by abutlerboudakian** [reply]
Some of that is copy-pasted from the library quickstart, but it routes correctly and generates a token, then stores it through the GoogleAuth model provided with the package, with hooks in the model code to auto-refresh when necessary

**Reply by TechStudent10** [reply]
@abutlerboudakian Ah okay. Impressive...

**Reply by TechStudent10** [reply]
I'm more of a Flask expert or guy rather than Django. But I guess it makes sense here.

**Comment by ThioJoe** [comment]
I'll first look into just hosting the Google Cloud project, where I'm imagining the user would still run the script locally and store the credentials on their own computer. Because security-wise, I'm _very_ hesitant to have to maintain custody of login credentials of users myself and host the actual script on my own site, so I would have to spend a lot of time making sure I do that properly.

But I think a good balance for now would be if no one had to go through the steps to set up their own cloud dashboard project and could just use one centralized one and I could request bigger quotas, but they'd still run the script locally, it would just call my API project instead of their own.

The one thing I have to look up is how they would be authorized to access the project. Right now obviously the user logs into the cloud dashboard and downloads their own client_secrets file, but I have to assume it's a different process for a public project.

**Reply by TechStudent10** [reply]
Yeah, I've seen different programs (locally from the command line) request a login to Heroku (for example) without a `client_secrets.json` file in the directory. It just logs in. Well, Heroku is probably using normal OAuth2 instead of whatever version of OAuth2 Google uses. But, it might work with some work.

**Reply by abutlerboudakian** [reply]
I've been thinking about the fork I've been working on for this and since you likely won't want to actually host a full fledged webapp version of this, and also regarding your point here about still wanting to have it so the end user doesn't have to create their own credentials.  Based on this, if I were to refactor what I have so far to get rid of all the frontend crap I've thrown together and just keep the Django project with the authentication implemented correctly, then added hooks in your CLI python file to automatically contact the Django server just for user authentication would that be something that would work for you in terms of eventually merging and making part of the standard application?

**Reply by abutlerboudakian** [reply]
The reason any hosting is needed in the first place for auth is so that the server hosting the API endpoints can store your Application level client_secrets,json file, which is your entrypoint for authentication, so if those are stored on the server users won't even need a copy of them to authenticate through your project.

**Reply by ArjixWasTaken** [reply]
But you can do that with a local auth server as well...

**Reply by abutlerboudakian** [reply]
Yes but a local auth server would still need to have access to the client_secrets.json to authenticate ThioJoe's central application, where a hosted solution for just auth allows those credentials to be moved to the server, with the delegated authentication to individual user accounts occurring locally without any static credential file needed for each end user. The flow is essentially ping the server with the client_secrets.json to retrieve a Google OAuth endpoint, then prompt the user to open the returned auth endpoint in their browser to perform the normal "Sign in with Google" process, then once that's done Google will perform a callback to the server which can then return the end-user's temporary access token and refresh token to the client application for the duration of client runtime.

**Reply by abutlerboudakian** [reply]
Doesn't necessarily need to be Django anymore at that point, honestly it probably wouldn't be the hardest thing in the world to do it in Flask as well since the remote server would just be hosting the authentication endpoints.

**Reply by ArjixWasTaken** [reply]
> Yes but a local auth server would still need to have access to the client_secrets.json to authenticate ThioJoe's central application, where a hosted solution for just auth allows those credentials to be moved to the server, with the delegated authentication to individual user accounts occurring locally without any static credential file needed for each end user. The flow is essentially ping the server with the client_secrets.json to retrieve a Google OAuth endpoint, then prompt the user to open the returned auth endpoint in their browser to perform the normal "Sign in with Google" process, then once that's done Google will perform a callback to the server which can then return the end-user's temporary access token and refresh token to the client application for the duration of client runtime.

Uhh, that makes no sense....there is no difference between a local and a remote server in this scenario.

**Reply by abutlerboudakian** [reply]
Can you clarify? How would a local server remove the need for the end user to have their own registered cloud app with their own client_secrets.json or a copy of ThioJoe's client_secrets.json?

**Reply by ArjixWasTaken** [reply]
> Can you clarify? How would a local server remove the need for the end user to have their own registered cloud app with their own client_secrets.json or a copy of ThioJoe's client_secrets.json?

Oh sorry for that, I did not account for google console app...
I had basic oath 2 login in my mind.

**Reply by abutlerboudakian** [reply]
Looking at this again, I think I understand the confusion here, when I'm referring to a hosted server, I'm not making the distinction of cloud vs running the application on ThioJoes machine or whoevers for the central auth service, I'm referring to the difference between having the client itself manage the authentication including the client_secrets.json, i.e. the end application that the end user uses, which is the current behavior, and having either a "local" or "hosted" server. Regardless of the means the goal it to have something that can be put somewhere other than the end users machine for authenticating. Apologies for the misunderstanding on my part.

**Reply by abutlerboudakian** [reply]
Oh wait just saw your latest reply, no worries. Right now I'm stuck trying to be able to get the Flask server (I'm using flask-dance now @TechStudent11 ) to properly sync up with both the main CLI with will be in the existing core application ThioJoe wrote, and the browser window that needs to be opened for the user to actually log into Google, as there are some state mismatches occurring with the current setup I have. I've used the starter project [Flask Dance Google Security SQLA](https://github.com/singingwolfboy/flask-dance-google-security-sqla) to get off the ground so far but I'm going to keep tweaking things today and hope I can get it to correctly begin the session through client-side CLI, contact the "remote" (localhost for now) Flask endpoint, have that return the actual Google OAuth url, then the user can open that, have the browser callback to Flask, then use a shared state to retrieve the user's access token back in the client-side CLI

**Reply by TechStudent10** [reply]
@abutlerboudakian That's awesome! Really hope everything works! Have you pushed the flask-dance version to your fork?

**Reply by abutlerboudakian** [reply]
I've pushed what I have so far, but it's a little hacky and still runs into the aforementioned issue I mentioned in my last reply, taking a break for now, will continue working on it either later today or tomorrow

**Reply by abutlerboudakian** [reply]
Everything's still in one file for the serverside code for the time being, will probably figure out how I want to redistribute all the code once I get everything working together finally.

**Reply by TechStudent10** [reply]
@abutlerboudakian You could use Flask Blueprints of something. Anyways, I'll check out your code. Do I have to provide my own `client_secrets.json` file?

**Reply by TechStudent10** [reply]
@abutlerboudakian After adding my own `client_secrets.json` file, and refactoring the config, the program gives me a URL. Apparently, I'll have to change some of my scopes to corporate. But so far so good!

**Reply by TechStudent10** [reply]
@abutlerboudakian how's it going so far?

**Reply by ThioJoe** [reply]
Another thing that needs to be considered is that for 'approval' from Google there is apparently quite an involved process. It would need to go through an audit, and because the app uses 'restricted' scopes (basically full access to the channel permissions) it will undergo the most scrutiny with the most requirements: https://support.google.com/cloud/answer/9110914?hl=en
As it stands now this is all avoided by having everyone just use their own personal project.

Edit: Actually looks like the youtube-force-ssl  scope is only a 'sensitive' scope, not restricted. Still it's more than trivial to get set up.

**Reply by TechStudent10** [reply]
@ThioJoe Aw shoot. First, Google makes getting a damn client ID and secret very complicated, and now getting approval from Google so that a self-hosted solution works even more complicated?! I take back everything I said about security and why Google makes these processes so complicated for the developer. If I wanted to add a "Sign In with Google" button to my website using Flask, I would not only have to get a client ID and secret but rewrite most of the Python 2.7 code that Google, for some reason, still uses on their site. Why Google, why?

Edit: Oh shoot I just pinged Google!

**Reply by abutlerboudakian** [reply]
Have not had time this week to look at this much, 9-5 took out most of my energy. Will have some time this weekend and next week.

**Reply by TechStudent10** [reply]
@abutlerboudakian alright! good luck!

**Reply by Kurdakov** [reply]
btw with community moderators feature which is scheduled to appear it might look like this: youtube channel owner grants community moderator rights to anti spam account after that the only thing he/she needs to provide is a link to their video for clean up

**Reply by TechStudent10** [reply]
@Kurdakov not a bad idea.

**Comment by KendallDoesCoding** [comment]
@ThioJoe this indeed would be cool, and much user-friendly. But as you mentioned it will take some research. For people that want to run it without downloading, in the browser any software though, I made a pull request #118 , it's mainly for unsupported devices, android users but it can work for anyone really. I just need @ThioJoe to check it out and to consider merging it in!
